fastly/openstack.go

```diff
@@ -0,0 +1,255 @@
+	// Service is the ID of the service. Version is the specific configuration
+	// version. Both fields are required.
+	Service string
+	Version int
+
+	Name              *string `form:"name,omitempty"`
+	User              *string `form:"user,omitempty"`
+	AccessKey         *string `form:"access_key,omitempty"`
+	BucketName        *string `form:"bucket_name,omitempty"`
+	URL               *string `form:"url,omitempty"`
+	Path              *string `form:"path,omitempty"`
+	Placement         *string `form:"placement,omitempty"`
+	Period            *uint   `form:"period,omitempty"`
+	GzipLevel         *uint   `form:"gzip_level,omitempty"`
+	Format            *string `form:"format,omitempty"`
+	FormatVersion     *uint   `form:"format_version,omitempty"`
+	ResponseCondition *string `form:"response_condition,omitempty"`
+	MessageType       *string `form:"message_type,omitempty"`
+	TimestampFormat   *string `form:"timestamp_format,omitempty"`
+	PublicKey         *string `form:"public_key,omitempty"`
+}
+
+// CreateOpenstack creates a new Fastly Openstack.
+func (c *Client) CreateOpenstack(i *CreateOpenstackInput) (*Openstack, error) {
+	if i.Service == "" {
+		return nil, ErrMissingService
+	}
+
+	if i.Version == 0 {
+		return nil, ErrMissingVersion
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/logging/openstack", i.Service, i.Version)
+	resp, err := c.PostForm(path, i, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var openstack *Openstack
+	if err := decodeBodyMap(resp.Body, &openstack); err != nil {
+		return nil, err
+	}
+	return openstack, nil
+}
+
+// GetOpenstackInput is used as input to the GetOpenstack function.
+type GetOpenstackInput struct {
+	// Service is the ID of the service. Version is the specific configuration
+	// version. Both fields are required.
+	Service string
+	Version int
+
+	// Name is the name of the Openstack to fetch.
+	Name string
+}
+
+// GetOpenstack gets the Openstack configuration with the given parameters.
+func (c *Client) GetOpenstack(i *GetOpenstackInput) (*Openstack, error) {
+	if i.Service == "" {
+		return nil, ErrMissingService
+	}
+
+	if i.Version == 0 {
+		return nil, ErrMissingVersion
+	}
+
+	if i.Name == "" {
+		return nil, ErrMissingName
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/logging/openstack/%s", i.Service, i.Version, url.PathEscape(i.Name))
+	resp, err := c.Get(path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var openstack *Openstack
+	if err := decodeBodyMap(resp.Body, &openstack); err != nil {
+		return nil, err
+	}
+	return openstack, nil
+}
+
+// UpdateOpenstackInput is used as input to the UpdateOpenstack function.
+type UpdateOpenstackInput struct {
+	// Service is the ID of the service. Version is the specific configuration
+	// version. Both fields are required.
+	Service string
+	Version int
+
+	// Name is the name of the Openstack to update.
+	Name string
+
+	NewName           *string `form:"name,omitempty"`
+	AccessKey         *string `form:"access_key,omitempty"`
```

Comment:
I'm going to dig deeper, but why isn't `User` here?

Reply:
I think this may have just been a typo. I don't see a reason for it to be missing.

Reply:
I just tried updating the field on manage.fastly.com and it is editable on updates. So I think it should be added.